EULER PROJECT

Exercise 1. If we list all the natural numbers below 10 that are multiples of 3 or 5, we get 3, 5, 6 and 9. The sum of these multiples is 23.

Find the sum of all the multiples of 3 or 5 below 1000.

In [3]:
def sumador(n):
    sum=0
    for i in range(1,n):

        if i%3 ==0 or i%5 == 0:
            sum = sum +i
    return(sum)

sumador(1000)

233168

Exercise 2. Each new term in the Fibonacci sequence is generated by adding the previous two terms. By starting with 1 and 2, the first 10 terms will be:

1, 2, 3, 5, 8, 13, 21, 34, 55, 89, ...

By considering the terms in the Fibonacci sequence whose values do not exceed four million, find the sum of the even-valued terms (terminos pares).

In [1]:
def fibonacci(n):
    
    num=[1,2]
    
    for i in range(2,n):
        

        num.append(num[i-2]+num[i-1])
         
    suma = 0
    
    for j in num:
        
        if j%2==0 and j<4000000:
            suma = suma+j
    return suma    


fibonacci(100)

4613732

Exercise 3. The prime factors of 13195 are 5, 7, 13 and 29.

What is the largest prime factor of the number 600851475143 ? Los numeros primos solo se pueden dividir entre uno o entre ellos mismos

In [265]:
import math 
  
def maxPrimeFactors2(n): 
    
    for i in range(3, int(math.sqrt(n)+1), 2): # Solo puede ir hasta la raiz. Es el limite maximo. Salta de 2 en dos para evitar pares
        if n % i == 0: 
            print('El numero ',i,'es candidato. Comprueba ahora si es primo')
            for m in range (3,i):
                if i % m == 0:
                    print('no es primo')
                    break
            else:
                print('es primo')


In [266]:
maxPrimeFactors2(600851475143)

El numero  71 es candidato. Comprueba ahora si es primo
es primo
El numero  839 es candidato. Comprueba ahora si es primo
es primo
El numero  1471 es candidato. Comprueba ahora si es primo
es primo
El numero  6857 es candidato. Comprueba ahora si es primo
es primo
El numero  59569 es candidato. Comprueba ahora si es primo
no es primo
El numero  104441 es candidato. Comprueba ahora si es primo
no es primo
El numero  486847 es candidato. Comprueba ahora si es primo
no es primo


Exercise 4. A palindromic number reads the same both ways. The largest palindrome made from the product of two 2-digit numbers is 9009 = 91 × 99.

Find the largest palindrome made from the product of two 3-digit numbers

In [35]:
import numpy as np

x = np.arange(1,1000)
y = np.arange(1,1000)


list1=[]

for i in range (0,998):
    for j in range (0,998):
        
        num=x[i]*y[j]
        num=num.astype(str)
        num_inv =(num[::-1])

        
        if num == num_inv:
            list1.append(num)
            
list1 = list(map(int, list1))
max(list1)

906609

Exercise 5. 2520 is the smallest number that can be divided by each of the numbers from 1 to 10 without any remainder.

What is the smallest positive number that is evenly divisible by all of the numbers from 1 to 20?

Hint: the number has to be a multiple of 2 3,4,...,10, and therefore it must be a multiple of the least common multiple of all of those, namely 2520. 

In [8]:
def func_miguel():

    i=2520

    while i > 2519:

        for j in range (11,22):     # from 1 to 10 it will be always divisible, so we start to count from 11

            if i % j != 0:
                break
            elif j == 21:
                return (i)

        i=i+2520    
        
func_miguel()

232792560

Exercise 6. Find the difference between the sum of the squares of the first one hundred natural numbers and the square of the sum.

In [81]:
import numpy as np

def difference():

    array = np.arange(1,101)


    for i in array:
        num1 = sum(array**2)
        num2 = sum(array)**2

        return abs(num1 - num2)

difference()    

25164150

Exercise 7. By listing the first six prime numbers: 2, 3, 5, 7, 11, and 13, we can see that the 6th prime is 13.

What is the 10 001st prime number? Los numeros primos solo se pueden dividir entre uno o entre ellos mismos

In [37]:
import math 

list2=[2,]

def funcion(n):

    i=1    

    while i>=1:

        for j in range (2,i):
            if i % j == 0:
#                print(i,'no es primo')
                break
            else:                
                if j==i-1:
#                    print(i,'es primo')
                    list2.append(i)
                    if len(list2)==n:
                        return()
                

        i=i+1
        
funcion(10001)
list2[10000]

# El algoritmo se podria hacer mas rapido descartando numeros pares

104743

Exercise 8. The four adjacent digits in the 1000-digit number that have the greatest product are 9 × 9 × 8 × 9 = 5832.

73167176531330624919225119674426574742355349194934
96983520312774506326239578318016984801869478851843
85861560789112949495459501737958331952853208805511
12540698747158523863050715693290963295227443043557
66896648950445244523161731856403098711121722383113
62229893423380308135336276614282806444486645238749
30358907296290491560440772390713810515859307960866
70172427121883998797908792274921901699720888093776
65727333001053367881220235421809751254540594752243
52584907711670556013604839586446706324415722155397
53697817977846174064955149290862569321978468622482
83972241375657056057490261407972968652414535100474
82166370484403199890008895243450658541227588666881
16427171479924442928230863465674813919123162824586
17866458359124566529476545682848912883142607690042
24219022671055626321111109370544217506941658960408
07198403850962455444362981230987879927244284909188
84580156166097919133875499200524063689912560717606
05886116467109405077541002256983155200055935729725
71636269561882670428252483600823257530420752963450

Find the thirteen adjacent digits in the 1000-digit number that have the greatest product. What is the value of this product?

In [223]:
import pandas as pd
import numpy as np

numbers='73167176531330624919225119674426574742355349194934\
96983520312774506326239578318016984801869478851843\
85861560789112949495459501737958331952853208805511\
12540698747158523863050715693290963295227443043557\
66896648950445244523161731856403098711121722383113\
62229893423380308135336276614282806444486645238749\
30358907296290491560440772390713810515859307960866\
70172427121883998797908792274921901699720888093776\
65727333001053367881220235421809751254540594752243\
52584907711670556013604839586446706324415722155397\
53697817977846174064955149290862569321978468622482\
83972241375657056057490261407972968652414535100474\
82166370484403199890008895243450658541227588666881\
16427171479924442928230863465674813919123162824586\
17866458359124566529476545682848912883142607690042\
24219022671055626321111109370544217506941658960408\
07198403850962455444362981230987879927244284909188\
84580156166097919133875499200524063689912560717606\
05886116467109405077541002256983155200055935729725\
71636269561882670428252483600823257530420752963450'

a=list(numbers)

In [225]:
nparray =np.asarray(a)

vector = nparray.reshape(1000).astype(int)

In [227]:
# limite inferior vector:

vector[0:13]

array([7, 3, 1, 6, 7, 1, 7, 6, 5, 3, 1, 3, 3])

In [228]:
# limite superior

vector[987:1000]

array([0, 4, 2, 0, 7, 5, 2, 9, 6, 3, 4, 5, 0])

In [229]:
import numpy as np

productos=[]

for i in range(0,988):
    prodvector = np.prod(vector[i:i+13])
    productos.append(prodvector)
    

max(productos)


23514624000

EJERCICIO 8 REALIZADO POR ERROR PENSANDO QUE HABIA QUE BUSCAR EL PRODUCTO DE MATRICES 13x13. MAS COMPLICADO DE LO QUE PIDE EL ENUNCIADO PERO PUEDE SERVIR PARA EL FUTURO:

In [3]:
import pandas as pd

numbers='73167176531330624919225119674426574742355349194934\
96983520312774506326239578318016984801869478851843\
85861560789112949495459501737958331952853208805511\
12540698747158523863050715693290963295227443043557\
66896648950445244523161731856403098711121722383113\
62229893423380308135336276614282806444486645238749\
30358907296290491560440772390713810515859307960866\
70172427121883998797908792274921901699720888093776\
65727333001053367881220235421809751254540594752243\
52584907711670556013604839586446706324415722155397\
53697817977846174064955149290862569321978468622482\
83972241375657056057490261407972968652414535100474\
82166370484403199890008895243450658541227588666881\
16427171479924442928230863465674813919123162824586\
17866458359124566529476545682848912883142607690042\
24219022671055626321111109370544217506941658960408\
07198403850962455444362981230987879927244284909188\
84580156166097919133875499200524063689912560717606\
05886116467109405077541002256983155200055935729725\
71636269561882670428252483600823257530420752963450'

In [4]:
mystring = str(numbers)

In [5]:
a=list(numbers)

In [6]:
len(a)

1000

In [7]:
type(a)

list

In [8]:
nparray =np.asarray(a)

matrix = nparray.reshape(20,50)

In [9]:
matrix.shape

(20, 50)

In [10]:
matrix.dtype

dtype('<U1')

In [11]:
matrix = matrix.astype(int)

In [12]:
matrix.dtype

dtype('int64')

In [13]:
matrix[19,49]

0

In [14]:
# matriz 13x13 superior izquierda

matrix[0:13,0:13]

array([[7, 3, 1, 6, 7, 1, 7, 6, 5, 3, 1, 3, 3],
       [9, 6, 9, 8, 3, 5, 2, 0, 3, 1, 2, 7, 7],
       [8, 5, 8, 6, 1, 5, 6, 0, 7, 8, 9, 1, 1],
       [1, 2, 5, 4, 0, 6, 9, 8, 7, 4, 7, 1, 5],
       [6, 6, 8, 9, 6, 6, 4, 8, 9, 5, 0, 4, 4],
       [6, 2, 2, 2, 9, 8, 9, 3, 4, 2, 3, 3, 8],
       [3, 0, 3, 5, 8, 9, 0, 7, 2, 9, 6, 2, 9],
       [7, 0, 1, 7, 2, 4, 2, 7, 1, 2, 1, 8, 8],
       [6, 5, 7, 2, 7, 3, 3, 3, 0, 0, 1, 0, 5],
       [5, 2, 5, 8, 4, 9, 0, 7, 7, 1, 1, 6, 7],
       [5, 3, 6, 9, 7, 8, 1, 7, 9, 7, 7, 8, 4],
       [8, 3, 9, 7, 2, 2, 4, 1, 3, 7, 5, 6, 5],
       [8, 2, 1, 6, 6, 3, 7, 0, 4, 8, 4, 4, 0]])

In [15]:
# matriz 13x13 inferior derecha

matrix[7:20,37:50]

array([[9, 7, 2, 0, 8, 8, 8, 0, 9, 3, 7, 7, 6],
       [4, 5, 4, 0, 5, 9, 4, 7, 5, 2, 2, 4, 3],
       [4, 4, 1, 5, 7, 2, 2, 1, 5, 5, 3, 9, 7],
       [1, 9, 7, 8, 4, 6, 8, 6, 2, 2, 4, 8, 2],
       [2, 4, 1, 4, 5, 3, 5, 1, 0, 0, 4, 7, 4],
       [1, 2, 2, 7, 5, 8, 8, 6, 6, 6, 8, 8, 1],
       [9, 1, 2, 3, 1, 6, 2, 8, 2, 4, 5, 8, 6],
       [3, 1, 4, 2, 6, 0, 7, 6, 9, 0, 0, 4, 2],
       [6, 9, 4, 1, 6, 5, 8, 9, 6, 0, 4, 0, 8],
       [7, 2, 4, 4, 2, 8, 4, 9, 0, 9, 1, 8, 8],
       [9, 9, 1, 2, 5, 6, 0, 7, 1, 7, 6, 0, 6],
       [0, 0, 5, 5, 9, 3, 5, 7, 2, 9, 7, 2, 5],
       [0, 4, 2, 0, 7, 5, 2, 9, 6, 3, 4, 5, 0]])

In [16]:
import numpy as np

for i in range (0,8):
    for j in range (0,38):

#        print ('i= (',i,':',i+13,') j= (',j,':',j+13,')')
        producto = np.prod(matrix[i:+13,j:j+13])
#        print ('producto = ',producto)
        
        
        

Problem 9. A Pythagorean triplet is a set of three natural numbers, a < b < c, for which,

a2 + b2 = c2
For example, 32 + 42 = 9 + 16 = 25 = 52.

There exists exactly one Pythagorean triplet for which a + b + c = 1000.
Find the product abc.

In [256]:
for a in range(0,999):

    for b in range(0,999):

        for c in range(0,999):
  
            
            if (a<b<c) and ((c**2) == (a**2) + (b**2)) and (a+b+c == 1000):
                
                print('a =',a,' b =',b,' c =',c, 'producto = ',a*b*c )
                
            else:
                continue
                    
    

a = 200  b = 375  c = 425 producto =  31875000


The sum of the primes below 10 is 2 + 3 + 5 + 7 = 17.

Find the sum of all the primes below two million.

In [176]:
def funcion(n):

    i=1    
    sum=2
    while i>=1:

        for j in range (2,i):
            if i % j == 0:
#                print(i,'no es primo')
                break
            else:                
                if j==i-1:
#                    print(i,'es primo')
                     sum=sum+i
                elif j>=n-1:
                    return(sum)
                

        i=i+1

# Este algoritmo para numeros muy grandes es demasiado lento

In [179]:
funcion(20)

77

In [166]:
# Algoritmo mas rapido encontrado en internet. Para comprobar si es primo, comprueba solo dividiendo hasta
# la raiz cuadrada del numero. 
# Para saber si un número es primo basta con probar si el número no es divisible por los primos hasta 
# su raiz cuadrada.

def is_prime(number):
    for i in range(2, int(number ** (1 / 2)) + 1):
        if number % i == 0:
            return False
    return True       

suma = 0

for i in range(2, 2 * 10 ** 6):
    if is_prime(i):
        suma=suma+i

print(suma)

142913828922


In the 20×20 grid below, four numbers along a diagonal line have been marked in red.

08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08
49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00
81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65
52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91
22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80
24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50
32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70
67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21
24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72
21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95
78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92
16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57
86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58
19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40
04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66
88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69
04 42 16 73 38 25 39 11 24 94 72 18 08 46 29 32 40 62 76 36
20 69 36 41 72 30 23 88 34 62 99 69 82 67 59 85 74 04 36 16
20 73 35 29 78 31 90 01 74 31 49 71 48 86 81 16 23 57 05 54
01 70 54 71 83 51 54 69 16 92 33 48 61 43 52 01 89 19 67 48

The product of these numbers is 26 × 63 × 78 × 14 = 1788696.

What is the greatest product of four adjacent numbers in the same direction (up, down, left, right, or diagonally) in the 20×20 grid?

In [3]:
import pandas as pd
import numpy as np

numbers='08 02 22 97 38 15 00 40 00 75 04 05 07 78 52 12 50 77 91 08 \
49 49 99 40 17 81 18 57 60 87 17 40 98 43 69 48 04 56 62 00 \
81 49 31 73 55 79 14 29 93 71 40 67 53 88 30 03 49 13 36 65 \
52 70 95 23 04 60 11 42 69 24 68 56 01 32 56 71 37 02 36 91 \
22 31 16 71 51 67 63 89 41 92 36 54 22 40 40 28 66 33 13 80 \
24 47 32 60 99 03 45 02 44 75 33 53 78 36 84 20 35 17 12 50 \
32 98 81 28 64 23 67 10 26 38 40 67 59 54 70 66 18 38 64 70 \
67 26 20 68 02 62 12 20 95 63 94 39 63 08 40 91 66 49 94 21 \
24 55 58 05 66 73 99 26 97 17 78 78 96 83 14 88 34 89 63 72 \
21 36 23 09 75 00 76 44 20 45 35 14 00 61 33 97 34 31 33 95 \
78 17 53 28 22 75 31 67 15 94 03 80 04 62 16 14 09 53 56 92 \
16 39 05 42 96 35 31 47 55 58 88 24 00 17 54 24 36 29 85 57 \
86 56 00 48 35 71 89 07 05 44 44 37 44 60 21 58 51 54 17 58 \
19 80 81 68 05 94 47 69 28 73 92 13 86 52 17 77 04 89 55 40 \
04 52 08 83 97 35 99 16 07 97 57 32 16 26 26 79 33 27 98 66 \
88 36 68 87 57 62 20 72 03 46 33 67 46 55 12 32 63 93 53 69 \
04 42 16 73 38 25 39 11 24 94 72 18 08 46 29 32 40 62 76 36 \
20 69 36 41 72 30 23 88 34 62 99 69 82 67 59 85 74 04 36 16 \
20 73 35 29 78 31 90 01 74 31 49 71 48 86 81 16 23 57 05 54 \
01 70 54 71 83 51 54 69 16 92 33 48 61 43 52 01 89 19 67 48'

In [4]:
type(numbers)

str

In [5]:
lista = list(numbers.split(" ")) 

In [6]:
type(lista)

list

In [10]:


matrix = np.asarray(lista).reshape(20,20).astype(int)

In [12]:
matrix.dtype

dtype('int64')

EXERCISE 12. The sequence of triangle numbers is generated by adding the natural numbers. So the 7th triangle number would be 1 + 2 + 3 + 4 + 5 + 6 + 7 = 28. The first ten terms would be:

1, 3, 6, 10, 15, 21, 28, 36, 45, 55, ...

Let us list the factors of the first seven triangle numbers:

 1: 1
 3: 1,3
 6: 1,2,3,6
10: 1,2,5,10
15: 1,3,5,15
21: 1,3,7,21
28: 1,2,4,7,14,28
We can see that 28 is the first triangle number to have over five divisors.

What is the value of the first triangle number to have over five hundred divisors?

In [170]:
# WITH "FOR" LOOP

def divisors(n):
    number_of_factors = 0
    for i in range(1, int(math.ceil(math.sqrt(n)))):
        if n % i == 0:
            number_of_factors +=2

    return number_of_factors

x=1
for y in range(2,1000000):
#    print (y)
    x += y   # si i+=1 significa que i vaya saltando de 1 en 1, x+=y significa que x vaya saltando de y en y
#    print('x=',x,'y=',y)
    if divisors(x) >= 500:
        
        print (x)
        break


76576500


In [171]:
#WITH "WHILE" LOOP

def divisors(n):
    number_of_factors = 0
    for i in range(1, int(math.ceil(math.sqrt(n)))):
        if n % i == 0:
            number_of_factors +=2

    return number_of_factors

x=0
y=1
while y>=1:
   
    x+=y
    y+=1
     
    if divisors(x) >= 500:
        
        print (x)
        break


76576500


PROBLEM 13. Work out the first ten digits of the sum of the following one-hundred 50-digit numbers.

In [22]:
import pandas as pd
import numpy as np

numbers="37107287533902102798797998220837590246510135740250\
46376937677490009712648124896970078050417018260538\
74324986199524741059474233309513058123726617309629\
91942213363574161572522430563301811072406154908250\
23067588207539346171171980310421047513778063246676\
89261670696623633820136378418383684178734361726757\
28112879812849979408065481931592621691275889832738\
44274228917432520321923589422876796487670272189318\
47451445736001306439091167216856844588711603153276\
70386486105843025439939619828917593665686757934951\
62176457141856560629502157223196586755079324193331\
64906352462741904929101432445813822663347944758178\
92575867718337217661963751590579239728245598838407\
58203565325359399008402633568948830189458628227828\
80181199384826282014278194139940567587151170094390\
35398664372827112653829987240784473053190104293586\
86515506006295864861532075273371959191420517255829\
71693888707715466499115593487603532921714970056938\
54370070576826684624621495650076471787294438377604\
53282654108756828443191190634694037855217779295145\
36123272525000296071075082563815656710885258350721\
45876576172410976447339110607218265236877223636045\
17423706905851860660448207621209813287860733969412\
81142660418086830619328460811191061556940512689692\
51934325451728388641918047049293215058642563049483\
62467221648435076201727918039944693004732956340691\
15732444386908125794514089057706229429197107928209\
55037687525678773091862540744969844508330393682126\
18336384825330154686196124348767681297534375946515\
80386287592878490201521685554828717201219257766954\
78182833757993103614740356856449095527097864797581\
16726320100436897842553539920931837441497806860984\
48403098129077791799088218795327364475675590848030\
87086987551392711854517078544161852424320693150332\
59959406895756536782107074926966537676326235447210\
69793950679652694742597709739166693763042633987085\
41052684708299085211399427365734116182760315001271\
65378607361501080857009149939512557028198746004375\
35829035317434717326932123578154982629742552737307\
94953759765105305946966067683156574377167401875275\
88902802571733229619176668713819931811048770190271\
25267680276078003013678680992525463401061632866526\
36270218540497705585629946580636237993140746255962\
24074486908231174977792365466257246923322810917141\
91430288197103288597806669760892938638285025333403\
34413065578016127815921815005561868836468420090470\
23053081172816430487623791969842487255036638784583\
11487696932154902810424020138335124462181441773470\
63783299490636259666498587618221225225512486764533\
67720186971698544312419572409913959008952310058822\
95548255300263520781532296796249481641953868218774\
76085327132285723110424803456124867697064507995236\
37774242535411291684276865538926205024910326572967\
23701913275725675285653248258265463092207058596522\
29798860272258331913126375147341994889534765745501\
18495701454879288984856827726077713721403798879715\
38298203783031473527721580348144513491373226651381\
34829543829199918180278916522431027392251122869539\
40957953066405232632538044100059654939159879593635\
29746152185502371307642255121183693803580388584903\
41698116222072977186158236678424689157993532961922\
62467957194401269043877107275048102390895523597457\
23189706772547915061505504953922979530901129967519\
86188088225875314529584099251203829009407770775672\
11306739708304724483816533873502340845647058077308\
82959174767140363198008187129011875491310547126581\
97623331044818386269515456334926366572897563400500\
42846280183517070527831839425882145521227251250327\
55121603546981200581762165212827652751691296897789\
32238195734329339946437501907836945765883352399886\
75506164965184775180738168837861091527357929701337\
62177842752192623401942399639168044983993173312731\
32924185707147349566916674687634660915035914677504\
99518671430235219628894890102423325116913619626622\
73267460800591547471830798392868535206946944540724\
76841822524674417161514036427982273348055556214818\
97142617910342598647204516893989422179826088076852\
87783646182799346313767754307809363333018982642090\
10848802521674670883215120185883543223812876952786\
71329612474782464538636993009049310363619763878039\
62184073572399794223406235393808339651327408011116\
66627891981488087797941876876144230030984490851411\
60661826293682836764744779239180335110989069790714\
85786944089552990653640447425576083659976645795096\
66024396409905389607120198219976047599490197230297\
64913982680032973156037120041377903785566085089252\
16730939319872750275468906903707539413042652315011\
94809377245048795150954100921645863754710598436791\
78639167021187492431995700641917969777599028300699\
15368713711936614952811305876380278410754449733078\
40789923115535562561142322423255033685442488917353\
44889911501440648020369068063960672322193204149535\
41503128880339536053299340368006977710650566631954\
81234880673210146739058568557934581403627822703280\
82616570773948327592232845941706525094512325230608\
22918802058777319719839450180888072429661980811197\
77158542502016545090413245809786882778948721859617\
72107838435069186155435662884062257473692284509516\
20849603980134001723930671666823555245252804609722\
53503534226472524250874054075591789781264330331690"

In [23]:
a=list(numbers)

In [24]:
matrix = np.array(a).reshape(100,50).astype(int)

In [25]:
l = []

suma_inici2 =0
for j in range (49,-1,-1):
    suma=0
    for i in range (0,100):
    
#        print(matrix[i,j])
        suma += matrix[i,j]
        
    suma2 = suma + suma_inici2
    suma2 = str(suma2)
    l.append(suma2[-1:])
    suma2 = int(suma2)
    suma_string = str(suma2)
#    print('suma =', suma_string)    
    suma_string2 = suma_string[-1:]
#    print('suma_ultimo_caracter = ', suma_string2)  
    
    suma_inici = suma_string[:-1]
    suma_inici2 = int(suma_inici)
    
#    print('primeros caracteres = ', suma_inici)
#    print('primeros caracteres = ', suma_inici2)
#    print('jump') 
#    print('nueva suma = ',suma2)

The first 10 digits of the sum are (5537376230)

PROBLEM 14 The following iterative sequence is defined for the set of positive integers:

n → n/2 (n is even)
n → 3n + 1 (n is odd)

Using the rule above and starting with 13, we generate the following sequence:

13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1
It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.

Which starting number, under one million, produces the longest chain?

NOTE: Once the chain starts the terms are allowed to go above one million.

In [14]:
import numpy as np

def collatz(n):
    array=[]
    array.append(n)
    i=0
    while i>=0:
        if array[i-1]==1:
            break
        if array[i-1] % 2 ==0: 
            array.append(int(array[i-1]/2))         
        else: 
            array.append(int((array[i-1]*3)+1))
        
    return array

In [34]:
list2=[]
list3=[]



for j in range (0,999999):
    list2.append(1000000-j)
    list2.append(len(collatz(1000000-j)))
    
    list3.append(len(collatz(1000000-j)))
    
file. close()


In [35]:
max(list3)

525

In [62]:
list2=[]
list3=[]

import os
file = open("/home/miguel/Desktop/lista.txt", "w")


for j in range (0,999999):
    
    file. write('\n')
    file. write('numero=')
    file. write(str(1000000-j))
    file. write('\n')
    file. write('longitud=')
    file. write(str(len(collatz(1000000-j))))
    file. write('\n')

    
file. close()

we check in the text file and these are the maximum numbers/length:

number=837799
length=525

PROBLEM 15. Starting in the top left corner of a 2×2 grid, and only being able to move to the right and down, there are exactly 6 routes to the bottom right corner.

How many such routes are there through a 20×20 grid?

In [19]:
results = {}

def routes(a, b):

    if (a, b) in results:
        return results[a, b]
#        print('a = ',a,'b= ',b, 'aqui copio de results')
    if a > b:
        r = routes(b, a)
#        print ('a = ',a,'b= ',b, 'aqui a>b')
    elif a == 0:
        r = 1
#        print ('a = ',a,'b= ',b, 'aqui a=0')
    else:
        r = routes(a - 1, b) + routes(a, b - 1)
#        print('a = ',a,'b= ',b, 'aqui calculo general')
    results[a, b] = r
    return r

In [20]:
routes(20,20)

137846528820

PROBLEM 16. 2**15 = 32768 and the sum of its digits is 3 + 2 + 7 + 6 + 8 = 26.

What is the sum of the digits of the number 2**1000?

In [22]:
a = 2**1000

In [23]:
a

10715086071862673209484250490600018105614048117055336074437503883703510511249361224931983788156958581275946729175531468251871452856923140435984577574698574803934567774824230985421074605062371141877954182153046474983581941267398767559165543946077062914571196477686542167660429831652624386837205668069376

In [24]:
type(a)

int

In [25]:
my_list = []
for i in str(a):
    my_list.append(i)

In [26]:
type(my_list)

list

In [28]:
type(my_list2)

list

In [27]:
sum(my_list2)

1366

PROBLEM 17. If the numbers 1 to 5 are written out in words: one, two, three, four, five, then there are 3 + 3 + 5 + 4 + 4 = 19 letters used in total.

If all the numbers from 1 to 1000 (one thousand) inclusive were written out in words, how many letters would be used?


NOTE: Do not count spaces or hyphens. For example, 342 (three hundred and forty-two) contains 23 letters and 115 (one hundred and fifteen) contains 20 letters. The use of "and" when writing out numbers is in compliance with British usage.

In [45]:
import inflect
p = inflect. engine()
words = p.number_to_words(1234)
print(words)

one thousand, two hundred and thirty-four


In [76]:
suma=0


for i in range (1,1001):
    
    
    words = p.number_to_words(i)
    words = words.replace(' ','')
    words = words.replace('-','')
#    print(words)
    word_total=len(words)
    suma=suma+word_total
print('suma =', suma)

suma = 21124
